data-pipeline-flink/notification-engine/src/main/scala/org/sunbird/dp/notification/function/IncompleteCourseReminderEmailNotification.scala

```diff
@@ -0,0 +1,298 @@
+import com.google.gson.Gson
+import org.apache.commons.collections.MapUtils
+import org.apache.commons.lang3.StringUtils
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord}
+import org.apache.kafka.common.serialization.StringSerializer
+import org.slf4j.LoggerFactory
+import org.sunbird.dp.core.util.CassandraUtil
+import org.sunbird.dp.notification.domain.Event
+import org.sunbird.dp.notification.task.NotificationEngineConfig
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+import java.util
+import java.util.{Date, Map, Properties, UUID}
+
+
+class IncompleteCourseReminderEmailNotification(courseConfig:NotificationEngineConfig)(implicit val mapTypeInfo: TypeInformation[Event]) {
+
+  case class CoursesDataMap(courseId: String, courseName: String, batchId: String, completionPercentage: Float, lastAccessedDate: java.util.Date, thumbnail: String, courseUrl: String, duration: String, description: String)
+
+  case class CourseDetails(courseName: String, thumbnail: String)
+
+  case class UserCourseProgressDetails(email: String=null, incompleteCourses: java.util.List[IncompleteCourse]=new util.ArrayList[IncompleteCourse]())
+
+  case class EmailConfig(sender: String, subject: String)
+
+  case class Notification(mode: String, deliveryType: String, config: EmailConfig, ids: java.util.List[String], template: Template)
+
+  case class Template(data: String, id: String, params: java.util.Map[String, Any])
+
+  case class IncompleteCourse(courseId: String, courseName: String,batchId:String,completionPercentage:Float,lastAccessedDate:Date,thumbnail:String,courseUrl:String)
+
+  var userCourseMap: java.util.Map[String, UserCourseProgressDetails] = new java.util.HashMap[String, UserCourseProgressDetails]()
+  val courseIdAndCourseNameMap: java.util.Map[String, CourseDetails] = new java.util.HashMap[String, CourseDetails]()
+
+
+  private[this] val logger = LoggerFactory.getLogger(classOf[IncompleteCourseReminderEmailNotification])
+
+  private var cassandraUtil = new CassandraUtil(courseConfig.dbHost, courseConfig.dbPort)
+
+  def initiateIncompleteCourseEmailReminder():Unit= {
+    try {
+      logger.info("Incomplete Course Reminder function started")
+      val date = new Date(new Date().getTime - courseConfig.last_access_time_gap_millis)
+      val query = QueryBuilder.select().all()
+        .from(courseConfig.dbCoursesKeyspace, courseConfig.USER_CONTENT_DB_TABLE).
+        where(QueryBuilder.gt("completionpercentage", 0))
+        .and(QueryBuilder.lt("completionpercentage", 100))
+        .and(QueryBuilder.gt("last_access_time", 0))
+        .and(QueryBuilder.lt("last_access_time", date))
+        .allowFiltering().toString
+      val rows: java.util.List[Row] = cassandraUtil.find(query)
+      if (rows != null) {
+        fetchCourseIdsAndSetCourseNameAndThumbnail(rows)
+        setUserCourseMap(rows, userCourseMap)
```

Comment:
so while executing the above code we get the details of course which is stored into userCourseMap and the same map used in the second method thats why I splited the functionality into two part

Reply:
This needs to be looked as part of future enhancements as currently the number of users may be limited, but when the user base increases, this can pose some delays. As discussed, marking this for later enhancement.
@Manas-tarento

@ Karthikeyan Rajendran - pls share your thoughts here

Reply:
Done In the latest PR #45